examples/light_sleep.rs

```diff
@@ -0,0 +1,164 @@
+use esp_idf_hal::prelude::*;
+use esp_idf_hal::reset::WakeupReason;
+use esp_idf_hal::sleep::*;
+use esp_idf_hal::uart::config::Config;
+use esp_idf_hal::uart::UartDriver;
+use std::thread;
+use std::time::Instant;
+
+use crate::gpio::Level;
+
+fn print_wakeup_result(time_before: Instant) {
+    let time_after = Instant::now();
+
+    let wakeup_reason = WakeupReason::get();
+    println!(
+        "wake up from light sleep due to {:?} which lasted for {:?}",
+        wakeup_reason,
+        time_after - time_before
+    );
+}
+
+fn main() -> anyhow::Result<()> {
+    esp_idf_sys::link_patches();
+
+    // run in a thread with increased stack size to prevent overflow
+    let builder = std::thread::Builder::new().stack_size(10 * 1024);
+    let th = builder.spawn(move || -> anyhow::Result<()> {
+        let peripherals = Peripherals::take().unwrap();
+
+        // RTC wakeup definitions
+        #[cfg(esp32)]
+        let rtc0 = PinDriver::rtc_input(peripherals.pins.gpio26)?;
+        #[cfg(esp32)]
+        let rtc1 = PinDriver::rtc_input(peripherals.pins.gpio27)?;
+
+        #[cfg(any(esp32s2, esp32s3))]
+        let rtc0 = PinDriver::rtc_input(peripherals.pins.gpio1)?;
+        #[cfg(any(esp32s2, esp32s3))]
+        let rtc1 = PinDriver::rtc_input(peripherals.pins.gpio2)?;
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        let rtc_pin0 = RtcWakeupPin { pindriver: &rtc0 };
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        let rtc_pin1 = RtcWakeupPin { pindriver: &rtc1 };
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        let rtc_wakeup = Some(RtcWakeup {
+            pins: EmptyRtcWakeupPins::chain(rtc_pin0).chain(rtc_pin1),
+            wake_level: RtcWakeLevel::AnyHigh,
+        });
+
+        // GPIO wakeup definitions
+        #[cfg(esp32)]
+        let gpio0 = PinDriver::input(peripherals.pins.gpio12)?;
+        #[cfg(esp32)]
+        let gpio1 = PinDriver::input(peripherals.pins.gpio14)?;
+
+        #[cfg(any(esp32s2, esp32s3))]
+        let gpio0 = PinDriver::input(peripherals.pins.gpio37)?;
+        #[cfg(any(esp32s2, esp32s3))]
+        let gpio1 = PinDriver::input(peripherals.pins.gpio38)?;
+
+        #[cfg(esp32c3)]
+        let gpio0 = PinDriver::input(peripherals.pins.gpio6)?;
+        #[cfg(esp32c3)]
+        let gpio1 = PinDriver::input(peripherals.pins.gpio7)?;
+
+        #[cfg(any(esp32, esp32c3, esp32s2, esp32s3))]
+        let gpio_pin0 = GpioWakeupPin {
+            pindriver: &gpio0,
+            wake_level: Level::High,
+        };
+        #[cfg(any(esp32, esp32c3, esp32s2, esp32s3))]
+        let gpio_pin1 = GpioWakeupPin {
+            pindriver: &gpio1,
+            wake_level: Level::High,
+        };
+        #[cfg(any(esp32, esp32c3, esp32s2, esp32s3))]
+        let gpio_wakeup = Some(GpioWakeup {
+            pins: EmptyGpioWakeupPins::chain(gpio_pin0).chain(gpio_pin1),
+        });
+        #[cfg(not(any(esp32, esp32c3, esp32s2, esp32s3)))]
+        let gpio_wakeup: Option<GpioWakeup> = None;
```

Comment:
If you don't want to use type annotation you can just use `let gpio_wakeup = None::<GpioWakeup>` so this a just a bit of style - like the other turbofish uses i mention down below